docs/architecture/blazor-for-web-forms-developers/components.md

```diff
@@ -3,11 +3,531 @@ title: Build reusable UI components with Blazor
+    {
+        <p><em>Loading...</em></p>
+    }
+    else
+    {
+        <p>The message is: @message</p>
+    }
+}
+
+@code 
+{
+    bool showMessage = false;
+    string message;
+
+    public async Task ShowMessage()
+    {
+        showMessage = true;
+        message = await MessageService.GetMessageAsync();
+    }
+}
+```
+
+Components can also define their own events by defining a component parameter of type `EventCallback<TValue>`. Event callbacks support all the variations of DOM UI event handlers: optional args, sync or async, method groups or lambda expressions.
+
+```razor
+<button class="btn btn-primary" @onclick="OnClick">Click me!</button>
+
+@code {
+    [Parameter]
+    public EventCallback<MouseEventArgs> OnClick { get; set; }
+}
+```
+
+## Data binding
+
+Blazor provides a simple mechanism to bind data from a UI component to the component's state. This is different than the various features in ASP.NET Web Forms for binding data from different data sources to UI controls. We'll cover handling data from different data sources in the [Dealing with data](data.md) section.
+
+To create a two-way data binding from a UI component to some the component's state use the `@bind` directive attribute. In the example below, the value of the check box is bound to the `isChecked` field.
+
+```razor
+<input type="checkbox" @bind="isChecked" />
+
+@code {
+    bool isChecked;
+}
+```
+
+When the component is rendered, the value of the checkbox will be set to the value of the `isChecked` field. When the user toggles the checkbox, the `onchange` is fired and the `isChecked` field is set to the new value. This means that the `@bind` syntax in this case is equivalent to the following.
+
+```razor
+<input value="@isChecked" @onchange="(UIChangeEventArgs e) => isChecked = e.Value" />
+```
+
+To change the event used for the bind, use the `@bind:event` attribute.
+
+```razor
+<input @bind="text" @bind:event="oninput" />
+<p>@text</p>
+
+@code {
+    string text;
+}
+```
+
+Components can also support data binding to their parameters by defining an event callback parameter with the same name as the bindable parameter, but with the "Changed" suffix added.
+
+*PasswordBox.razor*
+
+```razor
+Password: <input 
+    value="@Password" 
+    @oninput="OnPasswordChanged" 
+    type="@(showPassword ? "text" : "password")" />
+
+<label><input type="checkbox" @bind="showPassword" />Show password</label>
+
+@code {
+    private bool showPassword;
+
+    [Parameter]
+    public string Password { get; set; }
+
+    [Parameter]
+    public EventCallback<string> PasswordChanged { get; set; }
+
+    private Task OnPasswordChanged(ChangeEventArgs e)
+    {
+        Password = e.Value.ToString();
+        return PasswordChanged.InvokeAsync(Password);
+    }
+}
+```
+
+Note that to chain a data binding to an underlying UI element you'll need to set the value and handle the event directly on the UI element instead of using the `@bind` attribute.
```

Comment:
```suggestion
Note that to chain a data binding to an underlying UI element, you'll need to set the value and handle the event directly on the UI element instead of using the `@bind` attribute.
```